### PR TITLE
Use protocol buffers instead of custom XDR

### PR DESCRIPTION
**Pull request by calmh**
**Not done, not for merging, just for review and discussion.**

This changes our serialization code to use protocol buffers instead of our custom code. Todo/done:

- [X] Protocol structs should be protobuf
- [X] Database structs should be protobuf
- [x] Local discovery structs should be protobuf
by humans?
- [x] Clean up bitfields vs booleans (as per Audrius stuff)
- [x] Remove currently unused fields as these can be added painlessly in the future

Changes I think we should *not* do:

- The global discovery protocol. No need, might be nice to have human readable, no real reason to cause the disturbance.
- The relay protocol, as far as it concerns speaking *to* a relay. No reason to break it currently and leaving it unchanged lets us reap the benefits of the deployed infrastructure.

Some changes that were necessary but unexpected or non-obvious:

- ~~The protobuf serializer uses a `Size()` method which conflicted with our `Size` fields on `BlockInfo`. I've renamed it to `Length`.~~
- I've added a `Size` field instead of `CachedSize` to `FileInfo` and let this be serialized everywhere. It means we get it for free in the `TruncatedFileInfo` instead of having to extract it from the block list.
- ~~The `TruncatedFileInfo` type is "officialized" and moved into the protocol package, but it's still mostly used by the database. I'm not sure of this, maybe it can be undone and moved back to db as we have some protobufs stuff there as well (but didn't when I first made this change)...~~
- The version vector `type Vector []Counter` definition doesn't work with protobuf so I had to make it a struct which resulted in a lot of tedious changes everywhere, but it's conceptually the same.
- I added `FileName()` and `FileLength()` methods to the interface that covers both `FileInfo` and `FileInfoTruncated` so we can get the name or size of the interface without having to type assert to the underlying type.
- I threw out the conversion from 0.12 db:s. We'll need full rescan for this change, but on the other hand it ought to be the last time ever.

### CODE REVIEW

**Comment by AudriusButkevicius** [comment]
So it's sad that I spent a fair amount of time this aswell. We should sync up better next time.

**Comment by calmh** [comment]
Sorry

**Comment by calmh** [comment]
On the other hand maybe you did some smart things I didn't think of

**Comment by AudriusButkevicius** [comment]
The Size() issue can be worked around using sizer=false protosizer=true option.

This is too big to review on the phone but I'll peek at it this evening.

**Comment by AudriusButkevicius** [comment]
We should use proto3 syntax as thats 0 bytes on default value fields on the wire (such as cachedsize).
Also to make this easier to consume I propose we give up flags and go for bool IsDeleted etc.

I'll post my proto file this evening, which I think is superior to yours

**Comment by calmh** [comment]
I was only using proto2 as you argued for it last time we talked ;) So yes, lets do that.

**Comment by AudriusButkevicius** [comment]
I was an advocate of 2, but 3 is just around the corner (beta4 atm) and it has nice optimizations that are useful to us.

**Comment by calmh** [comment]
And a much cleaner schema language. All that "optional" and "nullable" crap is grating on me

**Comment by AudriusButkevicius** [comment]
You still need nullable. Just hang on until I post my schema and then we can fight it out whats write and whats wrong

**Comment by AudriusButkevicius** [comment]
My bep.proto
```proto
// Copyright (C) 2016 The Protocol Authors.

syntax = "proto3";

package protocol;

import "github.com/gogo/protobuf/gogoproto/gogo.proto";

option (gogoproto.goproto_getters_all) = false;
option (gogoproto.goproto_stringer_all) = true;
option (gogoproto.sizer_all) = false;
option (gogoproto.protosizer_all) = true;
option (gogoproto.goproto_enum_stringer_all) = true;
option (gogoproto.goproto_enum_prefix_all) = false;

message HelloMessage {
	string device_name = 1;
	string client_name = 2;
	string client_version = 3;
}

message IndexMessage {
	string folder = 1;
	repeated FileInfo files = 2 [(gogoproto.nullable) = false];
}

message FileInfo {
	string name = 1;
	int64 modified = 2;
	FileInfoType type = 3;
	uint32 permissions = 4;
	bool no_permissions = 5;
	bool deleted = 6;
	bool invalid = 7;
	Vector version = 8 [(gogoproto.nullable) = false];
	int64 local_version = 9;
	int64 cached_size = 10;
	repeated BlockInfo blocks = 11 [(gogoproto.nullable) = false];
}

enum FileInfoType {
	FILE					= 0 [(gogoproto.enumvalue_customname) = "FileInfoTypeFile"];
	DIRECTORY				= 1 [(gogoproto.enumvalue_customname) = "FileInfoTypeDirectory"];
	SYMLINK_FILE			= 2 [(gogoproto.enumvalue_customname) = "FileInfoTypeSymlinkFile"];
	SYMLINK_DIRECTORY		= 3 [(gogoproto.enumvalue_customname) = "FileInfoTypeSymlinkDirectory"];
	SYMLINK_UNKNOWN			= 4 [(gogoproto.enumvalue_customname) = "FileInfoTypeSymlinkUnknown"];
}

message BlockInfo {
	int64 offset = 1;
	int32 size = 2;
	bytes hash = 3;
}

message RequestMessage {
	string folder = 1;
	string name = 2;
	int64 offset = 3;
	int32 size = 4;
	bytes hash = 5;
	bool from_temporary = 6;
}

message ResponseMessage {
	bytes data = 1;
	ErrorCode code = 2;
}

enum ErrorCode {
	NO_ERROR 				= 0 [(gogoproto.enumvalue_customname) = "ErrorCodeNoError"];
	GENERIC					= 1 [(gogoproto.enumvalue_customname) = "ErrorCodeGeneric"];
	NO_SUCH_FILE			= 2 [(gogoproto.enumvalue_customname) = "ErrorCodeNoSuchFile"];
	INVALID_FILE			= 3 [(gogoproto.enumvalue_customname) = "ErrorCodeInvalidFile"];
}

message ClusterConfigMessage {
	repeated Folder folders = 1 [(gogoproto.nullable) = false];
}

message Folder {
	string id = 1;
	string label = 2;
	repeated Device devices = 3 [(gogoproto.nullable) = false];
	bool read_only = 4;
	bool ignore_permissions = 5;
	bool ignore_delete = 6;
	bool disable_temp_indexes = 7;
}

message Device {
	bytes id = 1;
	string name = 2;
	repeated string addresses = 3;
	uint32 compression = 4;
	string cert_name = 5;
	int64 max_local_version = 6;
	uint32 flags = 7;
	bool introducer = 8;
}

message CloseMessage {
	string reason = 1;
}

message EmptyMessage{
}

message PingMessage{
}

message FileDownloadProgressUpdate {
	FileDownloadProgressUpdateType update_type = 1;
	string name = 2;
	Vector version = 3 [(gogoproto.nullable) = false];
	repeated int32 block_indexes = 4;
}

enum FileDownloadProgressUpdateType {
	UPDATE					= 0 [(gogoproto.enumvalue_customname) = "FileDownloadProgressUpdateTypeUpdate"];
	FORGET 					= 1 [(gogoproto.enumvalue_customname) = "FileDownloadProgressUpdateTypeForget"];
}

message DownloadProgressMessage {
	string folder = 1;
	repeated FileDownloadProgressUpdate updates = 2 [(gogoproto.nullable) = false];
}

// The Vector type represents a version vector. The zero counter value is a usable
// version vector. The vector has slice semantics and some operations on it
// are "append-like" in that they may return the same vector modified, or v
// new allocated Vector with the modified contents.
message Vector {
	repeated Counter counters = 1 [(gogoproto.nullable) = false];
}

// Counter represents a single counter in the version vector.
message Counter {
	uint64 short_id = 1 [(gogoproto.customtype) = "ShortID", (gogoproto.nullable) = false, (gogoproto.customname) = "ID"];
	uint64 value = 2;
}
```

bep_extensions.go
```go
// Copyright (C) 2014 The Protocol Authors.

//go:generate protoc -I . -I ../../../../../ -I ../../../../../github.com/gogo/protobuf/protobuf --gogofaster_out=. bep.proto

package protocol

import (
	"bytes"
	"crypto/sha256"
)

var (
	sha256OfEmptyBlock        = sha256.Sum256(make([]byte, BlockSize))
	HelloMessageMagic  uint32 = 0x9F79BC40
)

func (f FileInfo) Size() (bytes int64) {
	if f.Deleted || f.IsDirectory() {
		return 128
	}
	if f.CachedSize > 0 {
		return f.CachedSize
	}
	for _, b := range f.Blocks {
		bytes += int64(b.Size)
	}
	f.CachedSize = bytes
	return
}

func (f FileInfo) IsDeleted() bool {
	return f.Deleted
}

func (f FileInfo) IsInvalid() bool {
	return f.Invalid
}

func (f FileInfo) IsDirectory() bool {
	return f.Type == FileInfoTypeDirectory
}

func (f FileInfo) IsSymlink() bool {
	return f.Type == FileInfoTypeSymlinkFile || f.Type == FileInfoTypeSymlinkDirectory || f.Type == FileInfoTypeSymlinkUnknown
}

func (f FileInfo) HasPermissionBits() bool {
	return !f.NoPermissions
}

// WinsConflict returns true if "f" is the one to choose when it is in
// conflict with "other".
func (f FileInfo) WinsConflict(other FileInfo) bool {
	// If a modification is in conflict with a delete, we pick the
	// modification.
	if !f.Deleted && other.IsDeleted() {
		return true
	}
	if f.Deleted && !other.IsDeleted() {
		return false
	}

	// The one with the newer modification time wins.
	if f.Modified > other.Modified {
		return true
	}
	if f.Modified < other.Modified {
		return false
	}

	// The modification times were equal. Use the device ID in the version
	// vector as tie breaker.
	return f.Version.Compare(other.Version) == ConcurrentGreater
}

// IsEmpty returns true if the block is a full block of zeroes.
func (b BlockInfo) IsEmpty() bool {
	return b.Size == BlockSize && bytes.Equal(b.Hash, sha256OfEmptyBlock[:])
}
```

**Comment by AudriusButkevicius** [comment]
Most significant changes:
1. Enums
2. Scrapping flags and options that are unused.
3. Expanding flags and options into booleans
4. Enum for FileInfo type
5. Uses ShortID type for Counter
6. Uses go generate

**Comment by calmh** [comment]
#1-6 I'll adopt directly (although I tried with ShortID as you do and couldn't get it to build, perhaps it works now in proto3).

#7 I thought is a good time to rectify by keeping the actual file size around. 

I think we don't need the emptymessage given we gain an explicit pingmessage, and perhaps the close message should gain an error code for symmetry since we have it in there?

Style wise I think I read somewhere that recommended for proto files is UpperCamel although it anyway becomes that in Go

**Comment by AudriusButkevicius** [comment]
https://developers.google.com/protocol-buffers/docs/style

UpperCamel for messages, snake_case for fields

**Comment by calmh** [comment]
Right you are.

Changing the fields and method signatures is rather tedious :D

File: lib/protocol/message.go [hunk 1]
```diff
@@ -17,28 +17,33 @@ var (
 )
 
 func (f FileInfo) String() string {
-	return fmt.Sprintf("File{Name:%q, Flags:0%o, Modified:%d, Version:%v, Length:%d, Blocks:%v}",
-		f.Name, f.Flags, f.Modified, f.Version, f.Size, f.Blocks)
+	return fmt.Sprintf("File{Name:%q, Permissions:0%o, Modified:%d, Version:%v, Length:%d, Deleted:%v, Invalid:%v, NoPermissions:%v, Blocks:%v}",
+		f.Name, f.Permissions, f.Modified, f.Version, f.Size, f.Deleted, f.Invalid, f.NoPermissions, f.Blocks)
 }
```
Comment (AudriusButkevicius) [inline lib/protocol/message.go]: Why just not proto.Striner?

**Comment by AudriusButkevicius** [comment]
UPPER_SNAKE for enums plz. Thats why the custom names are there to start with.

**Comment by calmh** [comment]
Yep fixed in the snake_case commit

**Comment by AudriusButkevicius** [comment]
Nvm.

File: lib/protocol/message.proto [hunk 2]
```diff
@@ -110,11 +110,14 @@ message DownloadProgressMessage {
...
 
 message Folder {
-    string          id      = 1 [(gogoproto.customname) = "ID"];
-    string          label   = 2;
-    repeated Device devices = 3 [(gogoproto.nullable) = false];
-    uint32          flags   = 4;
-    repeated Option options = 5 [(gogoproto.nullable) = false];
+    string id                   = 1 [(gogoproto.customname) = "ID"];
+    string label                = 2;
+    bool   read_only            = 3;
```
Comment (calmh) [inline lib/protocol/message.proto]: @AudriusButkevicius you recently made this a "folder type" enum internally, do we want to follow suite here?

**Comment by calmh** [comment]
FWIW I'm running this in "production" and it works. No real impact on the benchmarks I've ran either.

**Comment by AudriusButkevicius** [comment]
Made some last generic comments on the review

**Comment by capi** [comment]
I think this is awesome and I'm really looking forward to hitting production with this and use it from Java :+1:

**Comment by calmh** [comment]
I've requested some extra testing on this, as I think it's in principle mergeable.

**Comment by AudriusButkevicius** [comment]
I think some parts of my remarks as a general comment on the review are still not addressed.

**Comment by calmh** [comment]
Then you need to answer my notes there because I've marked all as done :)

**Comment by AudriusButkevicius** [comment]
Right,  I am travelling until next week, but I'll see if I can read it on the mobile.

**Comment by calmh** [comment]
No worries; there's no rush on this really. I don't mind letting it stew for a while and running the PR version of the code in the meantime. It's a little on the early side for a 0.14 break anyway.

**Comment by calmh** [comment]
@st-jenkins retest this please to see how the metalinter thing fares

**Comment by AudriusButkevicius** [comment]
So last nitpick on revews.st.net, and looks good to go.

**Comment by AudriusButkevicius** [comment]
@st-review lgtm

**Comment by st-review** [comment]
@AudriusButkevicius: Noted! Need another LGTM or explicit merge command.

**Comment by calmh** [comment]
@st-jenkins retest this please

I've increased the memory limits

**Comment by calmh** [comment]
I'm taking the opportunity here to reorder the message types logically as well, and moving the definition of the message type numbers in to the protocol buffer schema (even though the header isn't in protobuf format).

I'm thinking about slightly redesigning the header format while we're still at it though. It is currently:

```
     0                   1                   2                   3
     0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
    |  Ver  |       Message ID      |      Type     |   Reserved  |C|
    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
    |                            Length                             |
    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

but we might as well make it

```
     0                   1                   2                   3
     0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
    |                   Reserved                  |C|      Type     |
    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
    |                            Length                             |
    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

That is, just a type (8 bits) and a compression bit. There should be no need for versioning bits after the initial handshake. The message id field is only relevant for request/response and we can include that mechanism in the message itself instead?

**Comment by AudriusButkevicius** [comment]
To be honest, we could make the header protobuf too, which with varint stuff might be even shorter. Most of the protobuf code I used had this envelope type which essentially acted as a header/container.

**Comment by calmh** [comment]
You mean an über-message containing any of the other types, preceded by just the length word and the compression bit? because we need those two as well at the very least

**Comment by AudriusButkevicius** [comment]
Something like:

message Header
  int type = 1
  bool compressed = 2


Optionally putting the

**Comment by calmh** [comment]
```
     0                   1                   2                   3
     0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
    |C|                          Length                             |
    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

```
message Header {
    int32 nextMessageLength = 1;
    bool compressed = 2;
   int32 uncompressedLength = 3; // set when compressed=true
}
```

```
message Message {
    ClusterConfigMessage    cluster_config    = 1;
    IndexMessage            index             = 2;
    IndexMessage            index_update      = 3;
    RequestMessage          request           = 4;
    ResponseMessage         response          = 5;
    DownloadProgressMessage download_progress = 6;
    PingMessage             ping              = 7;
    CloseMessage            close             = 8;
}
```

**Comment by AudriusButkevicius** [comment]
Compression could also be an enum indicating type

**Comment by AudriusButkevicius** [comment]
Gzip bz xz zip etc

**Comment by calmh** [comment]
Yes

**Comment by calmh** [comment]
Okay... I went all on in the wrapper message thing. The structure is now:

```
     0                   1                   2                   3
     0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
    |                            Length                             |
    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
    /                                                               /
    \                            Message                            \
    /                                                               /
    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+

    message Message {
        ClusterConfig    cluster_config    = 1;
        Index            index             = 2;
        Index            index_update      = 3;
        Request          request           = 4;
        Response         response          = 5;
        DownloadProgress download_progress = 6;
        Ping             ping              = 7;
        Close            close             = 8;

        CompressedMessage compressed = 16;
    }

    message CompressedMessage {
        bytes data                = 1;
        int32 uncompressed_length = 2;
    }
```

When we use compression the message is marshalled, LZ4 compressed and stuffed into a CompressedMessage, which is then marshalled and sent over the wire. This way we need no separate header message and things become a bit easier in some respects. The "marshalling" step is essentially just a byte copy to a pre-allocated buffer just like it was anyway.

**Comment by AudriusButkevicius** [comment]
So I find this a bit weird to be honest.
Why was having a header like before a problem?

**Comment by calmh** [comment]
It was somewhat annoying to serialize as everything is variable size and need to be concatenated together. This way it's mostly one thing to marshal/unmarshal and the alternating length/message/length/message order is simple to understand and handle. I was thinking this would make more sense to people used to protocol buffers and nothing else, but maybe that's wrong?

**Comment by AudriusButkevicius** [comment]
So the problem is being unable to define where the header ends? How do you define where this nested message ends?

**Comment by calmh** [comment]
No no, it's not a problem, the sizes are all computable using Protosize, but it requires some buffer juggling and feels a bit unnecessary to have what is essentially a linked list of length fields when the only value added by the header message is a compression bit. 

With the nested setup above, we just write the length of the outer message and that's it. If there is a nested compressed message we get that as a []byte directly in the protobuf unmarshalled Message, so we run decompress+another unmarshal on that. 

I could see that part being a bit unusual though. Some more peer review on the design here would be good.

**Comment by calmh** [comment]
The lz4 compression also seems to only gain us about 20% on the protobuf marshaled index messages so it might be questionable if it's even worth it. But on the other hand it's almost free cpu wise.

**Comment by AudriusButkevicius** [comment]
I definately don't like this compound message business with a bunch of nulls. I'd rather have a nornal header like before.

**Comment by calmh** [comment]
Okay, that's fair. I had interpreted it that this was the common practice when doing protocol buffers, but lets look around and see what's what? You know where I come from so certainly a header makes sense to me as well. :)

**Comment by AudriusButkevicius** [comment]
I guess what I was refering to was an envelope, similar to what you have now, but instead of embedding unmarshaled messages , embed a byte string which holds the content.

**Comment by calmh** [comment]
Like what I did with the `compressed` thing?

**Comment by calmh** [comment]
https://forum.syncthing.net/t/interrobang-talk-to-me-about-protocol-buffers/7658

**Comment by AudriusButkevicius** [comment]
Hmm maybe, but just feels fugly to me.

**Comment by calmh** [comment]
I don't necessarily disagree.

**Comment by calmh** [comment]
Okay so that didn't result in as much input as I'd hoped, even with spreading the link on Twitter... Then I'd probably like to propose something closer to our original approach, basically sending:

 - uint8 `message_type` (we can have 256 message types)
 - uint8 `flags` (currently only one flag available, `lz4_compressed`)
 - uint32 `message_length`
 - uint32 `uncompressed_length` (when `lz4_compressed=1`)
 - ... the message in protobuf format

I don't want to include the serialized message in another protobuf by default as it's almost guaranteed to result in an unnecessary buffer copy + allocation per message.

We can make the above fairly flexible and future proof by simply mandating that clients silently ignore unknown messages with unknown type or flags - adding messages then becomes roughly the same as adding attributes inside a protobuf messages... You can do it, but must be prepared for the fact that older clients won't pick up on them.

**Comment by calmh** [comment]
This is now an "option C" implementation: four bytes of `header length` followed by a `Header`:

```proto
message Header {
    MessageType        type                = 1;
    int32              length              = 2;
    MessageCompression compression         = 3;
    int32              uncompressed_length = 4; // only set when compression != NONE
}

enum MessageCompression {
    NONE = 0 [(gogoproto.enumvalue_customname) = "MessageCompressionNone"];
    LZ4  = 1 [(gogoproto.enumvalue_customname) = "MessageCompressionLZ4"];
}
```

followed by one of the regular messages

**Comment by calmh** [comment]
@st-jenkins retest this please

**Comment by AudriusButkevicius** [comment]
Not sure why we need two length fields in the header? I am not even sure I can see it being used anywhere.

If you are worried about some expansion attack then whoever produces a malicious payload can also produce a malicious length value, and if we have a way to limit the amount we decompress before giving up, we might as well blow up once we are above 2*BlockSize or something along these lines.

**Comment by AudriusButkevicius** [comment]
Otherwise, merge whenever you feel it's done.

**Comment by calmh** [comment]
Yeah, no, the two fields are for compression. Or rather the first is just the regular "how much should I read" length field, the other is needed by the LZ4 package to be able to decompress the message. We used to stash that as an extra uint32 before the message, but I moved it into the header now. 

OK then, let me run some extra tests and stuff and update the docs, and then we can merge both together when it's good.

**Comment by AudriusButkevicius** [comment]
So let's just keep thr "hoe much to read" in length, and make decompressed length part of the payload (because lz4 expects it there no?) 

Not really sure why decompression needs to know the size ahead of time.

**Comment by calmh** [comment]
The LZ4 package *we* use expects it there, but in little endian byte order and I'm not sure how standard any of that is anyhow... It would be simpler for us to stash it there, sure, maybe converting it to big endian in the process to keep everything the same byte order.

File: lib/protocol/protocol.go [hunk 3]
```diff
@@ -421,6 +420,37 @@ func (c *rawConnection) readMessage() (message, error) {
 	return msg, nil
 }
 
+func (c *rawConnection) readHeader() (Header, error) {
+	// First comes a 2 byte header length
+
+	if len(c.readerBuf) < 2 {
```
Comment (AudriusButkevicius) [inline lib/protocol/protocol.go]: Why not cap()?

Comment (AudriusButkevicius) [inline lib/protocol/protocol.go]: To be honest, it should be cap() in a lot of other places in this package too?

Comment (AudriusButkevicius) [inline lib/protocol/protocol.go]: also, this whole 
```
if c.buf < y {
  c.buf = make()
}
use(c.buf)
```

should be which deals allocations and potentially resizing.

use(c.getBuf(123))

Comment (calmh) [inline lib/protocol/protocol.go]: We never shrink the readerbuf (and other buffers), so it'll start out empty on the first message, grow to like two bytes then to a message size and then possibly at some point to a bigger message. We always use it like  `Read(c.readerBuf[:totalSize])` and similar, so no need to futz around with `cap()` etc.

That said, we *should* probably do a `getBuf(123)` as you say and have a sync.Pool in the background. As it is it's fairly efficient in that we're doing zero allocations once up and running, but on the other hand each connection will have four or so buffers permanently allocated, each sized for the biggest message ever seen.

Comment (AudriusButkevicius) [inline lib/protocol/protocol.go]: If anything, the getBuf gets rid of multiple ifs.

**Comment by calmh** [comment]
It was actually cleaner code to use a buffer pool, regardless of the efficiency

**Comment by calmh** [comment]
I didn't worry about putting things back into the buffer pool in error conditions, as it's cleaner without it and the error conditions will result in the connection being torn down so we'll anyway have some garbage collection to do.

File: lib/protocol/protocol.go [hunk 4]
```diff
@@ -372,39 +368,35 @@ func (c *rawConnection) readMessage() (message, error) {
...
 		return nil, fmt.Errorf("negative message length %d", msgLen)
 	}
 
 	// Then comes the message
 
-	if len(c.readerBuf) < int(msgLen) {
-		c.readerBuf = make([]byte, msgLen)
-	}
-	if _, err := io.ReadFull(c.cr, c.readerBuf[:msgLen]); err != nil {
+	buf = buffers.upgrade(buf, int(msgLen))
```
Comment (AudriusButkevicius) [inline lib/protocol/protocol.go]: Why not put and get? I can't see the code for the pool implementation over the phone atleast.

Comment (calmh) [inline lib/protocol/protocol.go]: upgrade tries to just reslice it, and if it's too small it does put+get

**Comment by AudriusButkevicius** [comment]
You could just defer the puts no?

**Comment by calmh** [comment]
Not really, because the upgrades may give me the same or another buffer, and we switch out buf with the decompressed data when that happens. Not a big deal imho

**Comment by calmh** [comment]
I suggest this may be ready for merge now, unless there are some further concerns. There is still time for minor tweaks until the actual release in a couple of weeks.

**Comment by AudriusButkevicius** [comment]
Whats the rationale behind the port change?

**Comment by AudriusButkevicius** [comment]
That's the only thing I can pick on, it already has a LGTM from me.

**Comment by calmh** [comment]
So who wants to add another lgtm? :)

**Comment by AudriusButkevicius** [comment]
I think 22001 is not a good number. I usually use 22001 if I need to run two instances for example.

**Comment by calmh** [comment]
I reverted the port number change as it doesn't necessarily belong in this PR regardless of whether we should do it or not

**Comment by AudriusButkevicius** [comment]
You can add the second lgtm I think.

**Comment by calmh** [comment]
I can, but then it's cleaner if you just merge it

**Comment by AudriusButkevicius** [comment]
@st-review merge it

**Comment by st-review** [comment]
@AudriusButkevicius: Merge failed:

```
$ git fetch -f origin refs/pull/3276/head:pr-3276
$ git fetch -f origin master:orig/master
From github.com:syncthing/syncthing
   223a835..21f5b16  master     -> orig/master
$ git reset --hard
HEAD is now at 21f5b16 gui: Sort device folder lists by label
$ git checkout master
Already on 'master'
Your branch is up-to-date with 'origin/master'.
$ git reset --hard orig/master
HEAD is now at 21f5b16 gui: Sort device folder lists by label
$ git clean -fxd
Skipping repository syncthing/syncthing
Skipping repository syncthing/syncthing-android
Skipping repository syncthing/docs
Skipping repository syncthing/discosrv
Skipping repository syncthing/syncthing-gtk
Skipping repository syncthing/syncthing-inotify
$ git merge --squash --no-commit pr-3276
Removing lib/protocol/vector_xdr.go
Removing lib/protocol/vector_compare_test.go
Removing lib/protocol/vector_compare.go
Removing lib/protocol/message_xdr.go
Removing lib/protocol/message.go
Removing lib/protocol/header.go
Removing lib/protocol/fuzz_test.go
Removing lib/protocol/fuzz.go
Auto-merging lib/model/model_test.go
Auto-merging lib/model/model.go
CONFLICT (content): Merge conflict in lib/model/model.go
Removing lib/discover/localpackets_xdr.go
Removing lib/discover/localpackets.go
Auto-merging lib/discover/local.go
Removing lib/db/truncated.go
Removing lib/db/testdata/oldformat.db.zip
Removing lib/db/leveldb_xdr.go
Removing lib/db/leveldb_convert_test.go
Removing lib/db/leveldb_convert.go
Auto-merging lib/connections/service.go
Auto-merging cmd/syncthing/main.go
Auto-merging cmd/syncthing/gui.go
Squash commit -- not updating HEAD
Automatic merge failed; fix conflicts and then commit the result.

```

**Comment by AudriusButkevicius** [comment]
Welp needs rebase.

**Comment by calmh** [comment]
Right-o fixing

**Comment by AudriusButkevicius** [comment]
@st-review merge it

**Comment by st-review** [comment]
:ok_hand: Merged as fa0101bd60863b9094a223114f44c94dc0eaaeeb. Thanks, @calmh!

**Comment by calmh** [comment]
Holy fuck it's merged! \o/ Scary.

**Comment by AudriusButkevicius** [comment]
🎉 great job! Big achievement!

**Pull request closed without merging**